### PR TITLE
handling caps changes (resolution etc) in a stream

### LINKED ISSUE #27: segmentation fault on decoding files with varying resolution
After decoding h264 file that contains video with changing resolution, pipeline crashes with segmentation fault.
Probable cause: information about changed caps is not passed to the next element.
Example pipeline:
```elixir
use Membrane.Pipeline
  alias Membrane.H264.FFmpeg.{Parser, Decoder, Encoder}
  alias Membrane.File.{Sink, Source}

  @impl true
  def handle_init(path_to_file) do
    children = %{
      file: %Source{chunk_size: 40_960, location: path_to_file},
      parser: %Parser{framerate: {25, 1}},
      decoder: Decoder,
      encoder: Encoder,
      sink: %Sink{location: "output.h264"}
    }

    links = [
      link(:file)
      |> to(:parser)
      |> to(:decoder)
      |> to(:encoder)
      |> to(:sink)
    ]

    {{:ok, spec: %ParentSpec{children: children, links: links}}, %{}}
  end
```
Behaviour:
- segmentation fault when running above code
- no segmentation fault when decoding and saving directly to raw output
- segmentation fault when adding any component after decoder (e.g. scaler, encoder) - on component receiving frame with different size

### PR DESCRIPTION
**Pull request by agga1**
closes #27

### CODE REVIEW

File: c_src/membrane_h264_ffmpeg_plugin/parser.c [hunk 1]
```diff
@@ -69,21 +84,34 @@ UNIFEX_TERM parse(UnifexEnv *env, UnifexPayload *payload, State *state) {
 
   uint8_t *data_ptr = payload->data;
   size_t data_left = old_size;
-
-  while (data_left > 0) {
+  // TODO should we handle more than one context change?
```
Comment (mickel8) [inline c_src/membrane_h264_ffmpeg_plugin/parser.c:87]: Nice comment, appreciate it :+1:

File: c_src/membrane_h264_ffmpeg_plugin/parser.c [hunk 2]
```diff
@@ -83,6 +92,20 @@ UNIFEX_TERM parse(UnifexEnv *env, UnifexPayload *payload, State *state) {
...
+      if (resolution_changed(*last_res, state)) {
+        if (changes_cnt >= max_changes) {
+          max_changes *= 2;
+          changes = unifex_realloc(changes, max_changes * sizeof(resolution));
+        }
+
+        resolution new_res = {state->parser_ctx->width,
+                              state->parser_ctx->height, frames_cnt};
+        changes[changes_cnt] = new_res;
+        changes_cnt++;
```
Comment (daniel-jodlos) [inline c_src/membrane_h264_ffmpeg_plugin/parser.c:104]: I think it is would be more common to write it as: `changes[changes_cnt++] = new_res`. Not a problem of course, just a note

**Review by daniel-jodlos — APPROVED** [review]
I had really hard time finding issues in your code, good job. I left one some minor comments, mostly about naming, none of them really being any issue. Please remove that one block of commented out code before merging though.

File: lib/membrane_h264_ffmpeg/parser.ex [hunk 3]
```diff
@@ -249,8 +263,22 @@ defmodule Membrane.H264.FFmpeg.Parser do
...
+  #   %H264{
+  #     width: width,
+  #     height: height,
+  #     framerate: state.framerate,
+  #     alignment: state.alignment,
+  #     nalu_in_metadata?: state.attach_nalus?,
+  #     stream_format: :byte_stream,
+  #     profile: profile
+  #   }
+  # end
```
[the remark this answers is not in the document]
Comment (agga1) [reply lib/membrane_h264_ffmpeg/parser.ex:278]: My mistake, sorry

**Pull request merged**